integration-test/http/http_system_test.c

```diff
@@ -0,0 +1,369 @@
+    HTTPResponse_t response;
+    /* Represents header data that will be sent in an HTTP request. */
+    HTTPRequestHeaders_t requestHeaders;
+    /* Status returned by methods in HTTP Client Library API. */
+    HTTPStatus_t httpStatus;
+
+    assert( pMethod != NULL );
+    assert( pPath != NULL );
+
+    /* Initialize all HTTP Client library API structs to 0. */
+    ( void ) memset( &requestInfo, 0, sizeof( requestInfo ) );
+    ( void ) memset( &response, 0, sizeof( response ) );
+    ( void ) memset( &requestHeaders, 0, sizeof( requestHeaders ) );
+
+    /* Initialize the request object. */
+    requestInfo.pHost = SERVER_HOST;
+    requestInfo.hostLen = SERVER_HOST_LENGTH;
+    requestInfo.method = pMethod;
+    requestInfo.methodLen = strlen( pMethod );
+    requestInfo.pPath = pPath;
+    requestInfo.pathLen = strlen( pPath );
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. */
+    requestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Set the buffer used for storing request headers. */
+    requestHeaders.pBuffer = userBuffer;
+    requestHeaders.bufferLen = USER_BUFFER_LENGTH;
+
+    httpStatus = HTTPClient_InitializeRequestHeaders( &requestHeaders,
+                                                      &requestInfo );
+    TEST_ASSERT_EQUAL( HTTP_SUCCESS, httpStatus );
+    TEST_ASSERT_NOT_EQUAL( 0, requestHeaders.headersLen );
+
+    /* Initialize the response object. The same buffer used for storing
+     * request headers is reused here. */
+    response.pBuffer = userBuffer;
+    response.bufferLen = USER_BUFFER_LENGTH;
+
+    LogDebug( ( "Sending HTTP %.*s request to %.*s%.*s...",
+                ( int32_t ) requestInfo.methodLen, requestInfo.method,
+                ( int32_t ) SERVER_HOST_LENGTH, SERVER_HOST,
+                ( int32_t ) requestInfo.pathLen, requestInfo.pPath ) );
+    LogDebug( ( "Request Headers:\n%.*s\n"
+                "Request Body:\n%.*s\n",
+                ( int32_t ) requestHeaders.headersLen,
+                ( char * ) requestHeaders.pBuffer,
+                ( int32_t ) REQUEST_BODY_LENGTH, REQUEST_BODY ) );
+
+    /* Send request to HTTP server. */
+    httpStatus = HTTPClient_Send( pTransportInterface,
+                                  &requestHeaders,
+                                  ( uint8_t * ) REQUEST_BODY,
+                                  REQUEST_BODY_LENGTH,
+                                  &response,
+                                  0 );
+
+    /* If a network error is found, retry sending request for a count of MAX_RETRY_COUNT.
+     *  Proceed when MAX_RETRY_COUNT has been hit, or when a successful request is sent,
+     *  whichever comes first. */
+    if( ( httpStatus == HTTP_NETWORK_ERROR ) && ( retryCount < MAX_RETRY_COUNT ) )
+    {
+        LogDebug( ( "A network error has occured, retrying request." ) );
+        resetTest();
```

Comment:
If you reset the test does it jump to setUp()? If it does jump tp setUp, then won't this function be called again with retryCount set to 0?

Reply:
Oh I see that resetTest() just calls teardown then setup only.
```
void resetTest(void)
{
  tearDown();
  setUp();
}
```
so never mind.